runtime/commonMain/src/kotlinx/serialization/Coders.kt

```diff
@@ -125,31 +125,130 @@ interface Decoder {
 }
 
 interface CompositeDecoder {
-    val context: SerialModule
-    fun endStructure(desc: SerialDescriptor) {}
 
     /**
      * Results of [decodeElementIndex]
      */
     companion object {
         const val READ_DONE = -1
+
+        @Deprecated(
+            message = "READ_ALL cannot be longer returned by 'decodeElementIndex', use 'decodeSequentially' instead",
+            level = DeprecationLevel.WARNING
+        )
         const val READ_ALL = -2
         const val UNKNOWN_NAME = -3
     }
 
+    val context: SerialModule
+    val updateMode: UpdateMode
+
+    fun endStructure(desc: SerialDescriptor) {}
+
     /**
-     *  Returns either index or one of READ_XXX constants
+     * Checks whether the current decoder supports strictly ordered decoding of the data
+     * without calling to [decodeElementIndex].
+     * If the method returns `true`, the caller might skip [decodeElementIndex] calls
```

Comment:
`might` or `must`? I.e. can I still call `decodeElementIndex` even if `decodeSequentially` returned `true`?

Reply:
`might`, that's exactly the point for the potential DCE.
You can either avoid calling `decodeSequentially` at all or ignore its result

Reply:
This should've probably been given an name like `canDecodeSequenctially` and not that imperative name.
I'm implementing a BER sequence decoder and thought `decodeSequentially` would change some internal state but it looks like the result can be ignored.
Whilst it's still experimental, please rename this function to reflect this.

Reply:
Good point! I've created https://github.com/Kotlin/kotlinx.serialization/issues/1469